lib/updatehub/updatehub.c

```diff
@@ -0,0 +1,920 @@
+						api_header, strlen(api_header));
+		if (ret < 0) {
+			LOG_ERR("Unable add option to add updatehub header");
+			goto error;
+		}
+
+		ret = coap_packet_append_payload_marker(&request_packet);
+		if (ret < 0) {
+			LOG_ERR("Unable to append payload marker");
+			goto error;
+		}
+
+		ret = coap_packet_append_payload(&request_packet,
+						 &_ctx.payload,
+						 strlen(_ctx.payload));
+		if (ret < 0) {
+			LOG_ERR("Not able to append payload");
+			goto error;
+		}
+
+		break;
+
+	default:
+		LOG_ERR("Invalid method");
+		ret = -1;
+		goto error;
+	}
+
+	ret = send(sock, request_packet.data, request_packet.offset, 0);
+	if (ret < 0) {
+		LOG_ERR("Could not send request");
+		goto error;
+	}
+
+error:
+	k_free(data);
+
+	return ret;
+}
+
+static void install_update_cb(void)
+{
+	struct coap_packet response_packet;
+	u8_t *data = k_malloc(MAX_DOWNLOAD_DATA);
+
+	wait_fds();
+
+	int rcvd = recv(sock, data, MAX_DOWNLOAD_DATA, MSG_DONTWAIT);
+	if (rcvd <= 0) {
+		_ctx.code_status = UPDATEHUB_NETWORKING_ERROR;
+		LOG_ERR("Could not receive data");
+		goto cleanup;
+	}
+
+	if (coap_packet_parse(&response_packet, data, rcvd, NULL, 0) < 0) {
+		LOG_ERR("Invalid data received");
+		_ctx.code_status = UPDATEHUB_DOWNLOAD_ERROR;
+		goto cleanup;
+	}
+
+	_ctx.downloaded_size =
+		_ctx.downloaded_size +
+		(response_packet.max_len - response_packet.offset);
+
+	if (tc_sha256_update(
+		    &_ctx.sha256sum, response_packet.data + response_packet.offset,
+		    response_packet.max_len - response_packet.offset) < 1) {
+		LOG_ERR("Could not update sha256sum");
+		_ctx.code_status = UPDATEHUB_DOWNLOAD_ERROR;
+		goto cleanup;
+	}
+
+	if (flash_img_buffered_write(
+		    &_ctx.flash_ctx, response_packet.data + response_packet.offset,
+		    response_packet.max_len - response_packet.offset,
+		    _ctx.downloaded_size ==
+		    _ctx.block.total_size) < 0) {
+		LOG_ERR("Error to write on the flash");
+		_ctx.code_status = UPDATEHUB_INSTALL_ERROR;
+		goto cleanup;
+	}
+
+	if (coap_update_from_block(&response_packet, &_ctx.block) < 0) {
+		_ctx.code_status = UPDATEHUB_DOWNLOAD_ERROR;
+		goto cleanup;
+	}
+
+	if (coap_next_block(&response_packet, &_ctx.block) == 0) {
+		LOG_ERR("Could not get the next");
+		_ctx.code_status = UPDATEHUB_DOWNLOAD_ERROR;
+		goto cleanup;
+	}
+
+	if (_ctx.downloaded_size == _ctx.block.total_size) {
+		u8_t image_hash[TC_SHA256_DIGEST_SIZE];
+
+		if (tc_sha256_final(image_hash, &_ctx.sha256sum) < 1) {
+			LOG_ERR("Could not finish sha256sum");
+			_ctx.code_status = UPDATEHUB_DOWNLOAD_ERROR;
+			goto cleanup;
+		}
+
+		int i;
+		char buffer[2];
+		char sha256_image_dowloaded[SHA256SUM_STRING_SIZE];
+
+		memset(&sha256_image_dowloaded, 0, SHA256SUM_STRING_SIZE);
+		for (i = 0; i < TC_SHA256_DIGEST_SIZE; i++) {
+			snprintk(buffer, TC_SHA256_DIGEST_SIZE, "%02x",
+				 image_hash[i]);
+			strcat(&sha256_image_dowloaded[i], buffer);
+		}
+
+		if (strcmp(sha256_image_dowloaded, _update_info.sha256sum_image) !=
```

Comment:
The indentation looks a bit ugly here, what about splitting after the first parameter